File: app/src/main/java/com/might/dwan/cashcalendar/ui/fragments/UpdateProfileFragment.java
package com.might.dwan.cashcalendar.ui.fragments;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;

import com.might.dwan.cashcalendar.R;
import com.might.dwan.cashcalendar.data.db.DBManager;
import com.might.dwan.cashcalendar.data.db.db_models.UsersInfoDB;
import com.might.dwan.cashcalendar.data.manager.PreferencesManager;
import com.might.dwan.cashcalendar.data.models.UserModel;
import com.might.dwan.cashcalendar.ui.dialogs.PhotoDialog;
import com.might.dwan.cashcalendar.utils.ConstantManager;

/**
 * Created by Might on 19.09.2017.
 */

public class UpdateProfileFragment extends BaseFragment implements View.OnClickListener {

    private ImageView mAvatarImg;
    private EditText mNickNameEt, mNameEt, mSurnameEt;
    private Button mSaveBtn;

    private PhotoDialog mPhotoDialog;

    @Nullable @Override public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_update_profile, container, false);
        initViews(v);
        setupViews();
        return v;
    }

    private void initViews(View v) {
        mAvatarImg = (ImageView) v.findViewById(R.id.update_photo_img);
        mNickNameEt = (EditText) v.findViewById(R.id.update_nickname_et);
        mNameEt = (EditText) v.findViewById(R.id.update_name_et);
        mSurnameEt = (EditText) v.findViewById(R.id.update_surname_et);
        mSaveBtn = (Button) v.findViewById(R.id.update_save_btn);

        Toolbar toolbar = (Toolbar) v.findViewById(R.id.toolbar);
        toolbar.setTitle(R.string.title_profile);
        ((AppCompatActivity) getActivity()).setSupportActionBar(toolbar);
    }

    private void setupViews() {
        try {
            UsersInfoDB usersInfoDB = new UsersInfoDB();
            UserModel user = usersInfoDB.getUser(DBManager.get(getActivity()).getWritableDatabase(), PreferencesManager.get(getActivity()).getPreferences().getNickname());

            String user_nickname = user.getNickname();
            String user_name = user.getName();
            String user_surname = user.getSurname();
            if (user_nickname != null && !user_nickname.trim().equals("")) {
                mNickNameEt.setText(user_nickname);
            }

            if (user_name != null && !user_name.trim().equals("")) {
                mNameEt.setText(user_name);
            }

            if (user_surname != null && !user_surname.trim().equals("")) {
                mSurnameEt.setText(user_surname);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override public void onStart() {
        super.onStart();
        setListeners();
    }

    private void setListeners() {
        mSaveBtn.setOnClickListener(this);
        mAvatarImg.setOnClickListener(this);
    }

    @Override public void onClick(View v) {
        switch (v.getId()) {
            case R.id.update_save_btn:
                clickSave();
                break;
            case R.id.update_photo_img:
                openPhotoDialog();
                break;
        }
    }

    public void clickSave() {
        if (checkFields()) {
            updateProfile();
        }
    }

    private boolean checkFields() {
        if (!isValidEt(mNickNameEt)) {
            showToast(getString(R.string.profile_enter_nickname));
            return false;
        }
        if (!isValidEt(mNameEt)) {
            showToast(getString(R.string.profile_enter_name));
            return false;
        }

        if (!isValidEt(mSurnameEt)) {
            showToast(getString(R.string.profile_enter_surname));
            return false;
        }
        return true;
    }

    private boolean isValidEt(EditText editText) {
        return !editText.getText().toString().trim().equals("");
    }

    private void updateProfile() {
        try {
            UsersInfoDB usersInfoDB = new UsersInfoDB();
            usersInfoDB.addUser(DBManager.get(getActivity()).getReadableDatabase(), getTextFromEt(mNickNameEt), getTextFromEt(mNameEt), getTextFromEt(mSurnameEt));
            PreferencesManager.get(getActivity()).getPreferences().saveNickname(mNickNameEt.getText().toString().trim());
            getActivity().finish();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private String getTextFromEt(EditText editText) {
        return editText.getText().toString().trim();
    }

    private void openPhotoDialog() {
        mPhotoDialog = new PhotoDialog();
        mPhotoDialog.show(getFragmentManager(), "PhotoDialog");
    }

    @Override public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            case ConstantManager.REQUEST_GALLERY_PICK:
                getResultPickGallery(resultCode, data);
                break;
            case ConstantManager.REQUEST_CAMERA:
                getResultCamera(resultCode, data);
                break;
        }
    }

    private void getResultPickGallery(int resultCode, Intent data) {
        if (resultCode == Activity.RESULT_OK) {
            if (data == null) return;
            mAvatarImg.setImageURI(data.getData());
        }
    }

    private void getResultCamera(int resultCode, Intent data) {
        if(resultCode==Activity.RESULT_OK){
            if(data!=null){

            }
        }
    }
}

File: app/src/main/java/com/might/dwan/cashcalendar/ui/screens/update_profile/UpdateProfileFragment.kt
package com.might.dwan.cashcalendar.ui.screens.update_profile

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.widget.TextView
import com.might.dwan.cashcalendar.R
import com.might.dwan.cashcalendar.apps.App
import com.might.dwan.cashcalendar.data.db.DBManager
import com.might.dwan.cashcalendar.data.db.db_writer.UsersInfoDB
import com.might.dwan.cashcalendar.data.preferences.Preferences
import com.might.dwan.cashcalendar.ui.dialogs.PhotoDialog
import com.might.dwan.cashcalendar.ui.screens.BaseFragment
import com.might.dwan.cashcalendar.ui.screens.update_profile.dagger.DaggerUpdateProfileComponent
import com.might.dwan.cashcalendar.ui.screens.update_profile.dagger.UpdateProfileComponent
import com.might.dwan.cashcalendar.utils.BitmapUtils
import com.might.dwan.cashcalendar.utils.ConstantManager
import com.might.dwan.cashcalendar.utils.EditTextUtils
import com.might.dwan.cashcalendar.utils.ValidUtils
import kotlinx.android.synthetic.main.fragment_update_profile.*
import javax.inject.Inject

/**
 * Created by Might on 19.09.2017.
 */

class UpdateProfileFragment : BaseFragment(), View.OnClickListener {

    @Inject
    lateinit var preferences: Preferences

    private lateinit var component: UpdateProfileComponent

    override fun getLayoutId() = R.layout.fragment_update_profile

    override fun initUI(v: View) {
        toolbar.setTitle(R.string.title_profile)
        setToolbar(toolbar)
    }

    override fun setupData(state: Bundle?) {
        try {
            component = DaggerUpdateProfileComponent.builder()
                    .appComponent(App.getAppComponent())
                    .build()

            component.inject(this)

            val nickName = preferences.nickname
            val usersInfoDB = UsersInfoDB()
            val user = usersInfoDB.getUser(DBManager.get(activity).writableDatabase, nickName)

            setTextIfNeed(update_nickname_et, user.nickname)
            setTextIfNeed(update_name_et, user.name)
            setTextIfNeed(update_surname_et, user.surname)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    private fun setTextIfNeed(v: TextView, text: String) {
        if (ValidUtils.isTextValid(text)) {
            v.text = text
        }
    }

    override fun setListeners(enable: Boolean) {
        if (enable) {
            update_save_btn.setOnClickListener(this)
            update_photo_img.setOnClickListener(this)
        } else {
            update_save_btn.setOnClickListener(null)
            update_photo_img.setOnClickListener(null)
        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.update_save_btn -> clickSave()
            R.id.update_photo_img -> openPhotoDialog()
        }
    }

    private fun clickSave() {
        if (checkFields()) {
            updateProfile()
        }
    }

    private fun checkFields(): Boolean {
        if (!EditTextUtils.isValid(update_nickname_et)) {
            showToast(getString(R.string.profile_enter_nickname))
            return false
        }
        if (!EditTextUtils.isValid(update_name_et)) {
            showToast(getString(R.string.profile_enter_name))
            return false
        }

        if (!EditTextUtils.isValid(update_surname_et)) {
            showToast(getString(R.string.profile_enter_surname))
            return false
        }
        return true
    }

    private fun updateProfile() {
        try {
            val usersInfoDB = UsersInfoDB()
            usersInfoDB.addUser(DBManager.get(activity).readableDatabase,
                    EditTextUtils.getText(update_nickname_et),
                    EditTextUtils.getText(update_name_et),
                    EditTextUtils.getText(update_surname_et))
            preferences.saveNickname(EditTextUtils.getText(update_nickname_et))
            activity?.finish()
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    private fun openPhotoDialog() {
        val dialog = PhotoDialog()
        component.inject(dialog)
        dialog.show(fragmentManager, dialog.javaClass.simpleName)
    }


    //Activity result region
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)
        val result = resultCode == Activity.RESULT_OK
        when (requestCode) {
            ConstantManager.REQUEST_GALLERY_PICK -> getResultPickGallery(result, data)
            ConstantManager.REQUEST_CAMERA -> getResultCamera(result, data)
            else -> return
        }
    }

    private fun getResultPickGallery(result: Boolean, data: Intent?) {
        if (data == null || !result) return
        update_photo_img.setImageURI(data.data)
        setPhoto(data.dataString)
    }

    private fun getResultCamera(result: Boolean, data: Intent?) {
        if (data == null || !result) return
        setPhoto(preferences.photoPath)
    }

    private fun setPhoto(photoPath: String?) {
        if (!ValidUtils.isTextValid(photoPath)) return
        try {
            val first = BitmapUtils.getBitmapFromUri(activity, Uri.parse(photoPath))
            update_photo_img.setImageBitmap(first)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }
}
